embassy-stm32/src/pka/mod.rs

```diff
@@ -0,0 +1,485 @@
+    /// Arithmetic Subtraction
+    ArithmeticSubtraction = 0b001010,
+    /// Arithmetic Multiplication
+    ArithmeticMultiplication = 0b001011,
+    /// Arithmetic Comparison
+    ArithmeticComparison = 0b001100,
+    /// Modular Reduction
+    ModularReduction = 0b001101,
+    /// Modular Addition
+    ModularAddition = 0b001110,
+    /// Modular Subtraction
+    ModularSubtraction = 0b001111,
+    /// Montgomery Multiplication
+    MontgomeryMultiplication = 0b010000,
+}
+
+/// PKA SealedInstance trait.
+///
+/// Allows access to PAC exclusively for `PKA` module.
+trait SealedInstance {
+    fn regs() -> pac::pka::Pka;
+}
+
+/// PKA interrupt
+pub struct InterruptHandler<T: Instance> {
+    __phantom: PhantomData<T>,
+}
+
+impl<T: Instance> interrupt::typelevel::Handler<T::Interrupt> for InterruptHandler<T> {
+    unsafe fn on_interrupt() {
+        let procendf = T::regs().sr().read().procendf();
+        let addr_err = T::regs().sr().read().addrerrf();
+        let ram_err = T::regs().sr().read().ramerrf();
+
+        if procendf {
+            T::regs().cr().modify(|w| w.set_procendie(false));
+        }
+
+        if addr_err {
+            T::regs().cr().modify(|w| w.set_addrerrie(false));
+        }
+
+        if ram_err {
+            T::regs().cr().modify(|w| w.set_ramerrie(false));
+        }
+
+        PKA_WAKER.wake();
+    }
+}
+
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+/// RSA structure for PKA accelerator
+pub struct Rsa<'c, 'd, T: Instance, const RSA_KEY_SIZE: usize>
+where
+    [(); RSA_KEY_SIZE / 4]:,
+{
+    /// Modulus
+    pub modulus: [u8; RSA_KEY_SIZE],
+    /// R^2 mod N
+    r2_mod_n: [u8; RSA_KEY_SIZE],
+    /// public exponent
+    pub public_exponent: [u8; RSA_KEY_SIZE],
+    /// private exponent
+    /// FIXME: In some edge cases the private exponent can be larger than the modulus
+    pub private_exponent: [u8; RSA_KEY_SIZE],
+    /// PKA accelerator
+    pka: &'c mut Pka<'d, T, RSA_KEY_SIZE>,
+}
+
+impl<'c, 'd, T: Instance, const RSA_KEY_SIZE: usize> Rsa<'c, 'd, T, RSA_KEY_SIZE>
+where
+    [(); RSA_KEY_SIZE / 4]:,
+{
+    /// Create a new RSA structure
+    pub fn new(pka: &'c mut Pka<'d, T, RSA_KEY_SIZE>) -> Self {
+        pka.enable();
+        Self {
+            modulus: [0; RSA_KEY_SIZE],
+            r2_mod_n: [0; RSA_KEY_SIZE],
+            public_exponent: [0; RSA_KEY_SIZE],
+            private_exponent: [0; RSA_KEY_SIZE],
+            pka,
+        }
+    }
+
+    /// Reorder data for the new word order
+    fn word_reorder<const N: usize>(input: &[u8; N]) -> [u8; N] {
+        // Ensure the input length is a multiple of 4 by padding with 0s if necessary
+        assert!(N % 4 == 0, "Input length must be a multiple of 4");
+
+        let mut reordered = [0u8; N];
+
+        for i in 0..(N / 4) {
+            let src_index = (N / 4 - 1 - i) * 4;
+            let dst_index = i * 4;
+            reordered[dst_index..dst_index + 4].copy_from_slice(&input[src_index..src_index + 4]);
```

Comment:
If something goes wrong with the indexing,`copy_from_slice` will fail. Can we maybe gate the slice grabbing with [slice::get(_)](https://doc.rust-lang.org/std/primitive.slice.html#method.get). Then on the options returned we can convert to an `Error` type and avoid the panic.